File: lib/analysis/wav_reader.dart
import 'dart:typed_data';

/// Mono (or downmixed) PCM samples from a WAV byte buffer.
class WavPcm {
  const WavPcm({
    required this.samples,
    required this.sampleRateHz,
    required this.channels,
  });

  final Float64List samples;
  final int sampleRateHz;
  final int channels;
}

/// Parses PCM and WAVE_FORMAT_EXTENSIBLE WAVs (mpv `ao=pcm` output).
class WavReader {
  const WavReader();

  WavPcm read(Uint8List bytes) {
    if (bytes.length < 12) {
      throw const FormatException('WAV too short');
    }
    if (!_eq(bytes, 0, 'RIFF') || !_eq(bytes, 8, 'WAVE')) {
      throw const FormatException('not a RIFF/WAVE file');
    }

    var offset = 12;
    int? sampleRate;
    int? channels;
    int? bitsPerSample;
    int? audioFormat;
    Uint8List? data;

    while (offset + 8 <= bytes.length) {
      final id = String.fromCharCodes(bytes.sublist(offset, offset + 4));
      final size = ByteData.sublistView(bytes, offset + 4, offset + 8)
          .getUint32(0, Endian.little);
      final body = offset + 8;
      if (body + size > bytes.length) break;

      if (id == 'fmt ') {
        final view = ByteData.sublistView(bytes, body, body + size);
        audioFormat = view.getUint16(0, Endian.little);
        channels = view.getUint16(2, Endian.little);
        sampleRate = view.getUint32(4, Endian.little);
        bitsPerSample = view.getUint16(14, Endian.little);
        // WAVE_FORMAT_EXTENSIBLE: real format is at cbSize offset.
        if (audioFormat == 0xFFFE && size >= 40) {
          audioFormat = view.getUint16(24, Endian.little);
        }
      } else if (id == 'data') {
        data = bytes.sublist(body, body + size);
      }

      offset = body + size + (size.isOdd ? 1 : 0);
    }

    if (sampleRate == null ||
        channels == null ||
        bitsPerSample == null ||
        audioFormat == null ||
        data == null) {
      throw const FormatException('WAV missing fmt/data');
    }
    if (audioFormat != 1 && audioFormat != 3) {
      throw FormatException('unsupported WAV format $audioFormat');
    }
    if (bitsPerSample != 16 && bitsPerSample != 32 && bitsPerSample != 24) {
      throw FormatException('unsupported bits $bitsPerSample');
    }

    final mono = _decodeMono(
      data,
      channels: channels,
      bitsPerSample: bitsPerSample,
      float: audioFormat == 3,
    );
    return WavPcm(
      samples: mono,
      sampleRateHz: sampleRate,
      channels: channels,
    );
  }

  static Float64List _decodeMono(
    Uint8List data, {
    required int channels,
    required int bitsPerSample,
    required bool float,
  }) {
    final bytesPerSample = bitsPerSample ~/ 8;
    final frameSize = bytesPerSample * channels;
    if (frameSize == 0) return Float64List(0);
    final frames = data.length ~/ frameSize;
    final out = Float64List(frames);
    final view = ByteData.sublistView(data);

    for (var i = 0; i < frames; i++) {
      var sum = 0.0;
      for (var ch = 0; ch < channels; ch++) {
        final off = i * frameSize + ch * bytesPerSample;
        if (float && bitsPerSample == 32) {
          sum += view.getFloat32(off, Endian.little);
        } else if (bitsPerSample == 16) {
          sum += view.getInt16(off, Endian.little) / 32768.0;
        } else if (bitsPerSample == 24) {
          final b0 = data[off];
          final b1 = data[off + 1];
          final b2 = data[off + 2];
          var v = b0 | (b1 << 8) | (b2 << 16);
          if (v & 0x800000 != 0) v |= ~0xFFFFFF;
          sum += v / 8388608.0;
        } else if (bitsPerSample == 32) {
          sum += view.getInt32(off, Endian.little) / 2147483648.0;
        }
      }
      out[i] = sum / channels;
    }
    return out;
  }

  static bool _eq(Uint8List bytes, int offset, String ascii) {
    for (var i = 0; i < ascii.length; i++) {
      if (bytes[offset + i] != ascii.codeUnitAt(i)) return false;
    }
    return true;
  }
}

File: src/wav_reader.cpp
#include "wav_reader.h"

#include <QByteArray>
#include <QtEndian>
#include <cstring>
#include <stdexcept>

namespace tramp {
namespace {

quint32 u32le(const QByteArray& bytes, int offset) {
  return qFromLittleEndian<quint32>(reinterpret_cast<const uchar*>(bytes.constData() + offset));
}

quint16 u16le(const QByteArray& bytes, int offset) {
  return qFromLittleEndian<quint16>(reinterpret_cast<const uchar*>(bytes.constData() + offset));
}

qint32 i32le(const QByteArray& bytes, int offset) {
  return qint32(u32le(bytes, offset));
}

qint16 i16le(const QByteArray& bytes, int offset) {
  return qint16(u16le(bytes, offset));
}

float f32le(const QByteArray& bytes, int offset) {
  const quint32 bits = u32le(bytes, offset);
  float value = 0;
  static_assert(sizeof(float) == 4, "float must be 32-bit");
  memcpy(&value, &bits, 4);
  return value;
}

bool eq4(const QByteArray& bytes, int offset, const char* ascii) {
  return bytes.size() >= offset + 4 && bytes[offset] == ascii[0] && bytes[offset + 1] == ascii[1] &&
         bytes[offset + 2] == ascii[2] && bytes[offset + 3] == ascii[3];
}

QVector<double> decodeMono(const QByteArray& data, int channels, int bitsPerSample, bool isFloat) {
  const int bytesPerSample = bitsPerSample / 8;
  const int frameSize = bytesPerSample * channels;
  if (frameSize <= 0) return {};
  const int frames = data.size() / frameSize;
  QVector<double> out(frames);
  for (int i = 0; i < frames; ++i) {
    double sum = 0;
    for (int ch = 0; ch < channels; ++ch) {
      const int off = i * frameSize + ch * bytesPerSample;
      if (isFloat && bitsPerSample == 32) {
        sum += double(f32le(data, off));
      } else if (bitsPerSample == 16) {
        sum += double(i16le(data, off)) / 32768.0;
      } else if (bitsPerSample == 24) {
        const int b0 = uchar(data[off]);
        const int b1 = uchar(data[off + 1]);
        const int b2 = uchar(data[off + 2]);
        int v = b0 | (b1 << 8) | (b2 << 16);
        if (v & 0x800000) v |= ~0xFFFFFF;
        sum += double(v) / 8388608.0;
      } else if (bitsPerSample == 32) {
        sum += double(i32le(data, off)) / 2147483648.0;
      }
    }
    out[i] = sum / channels;
  }
  return out;
}

}  // namespace

PcmBuffer WavReader::read(const QByteArray& bytes) const {
  if (bytes.size() < 12 || !eq4(bytes, 0, "RIFF") || !eq4(bytes, 8, "WAVE")) {
    throw std::runtime_error("not a RIFF/WAVE file");
  }

  int offset = 12;
  int sampleRate = 0;
  int channels = 0;
  int bitsPerSample = 0;
  int audioFormat = 0;
  QByteArray data;
  bool haveFmt = false;
  bool haveData = false;

  while (offset + 8 <= bytes.size()) {
    const int size = int(u32le(bytes, offset + 4));
    const int body = offset + 8;
    if (body + size > bytes.size()) break;

    if (eq4(bytes, offset, "fmt ")) {
      audioFormat = u16le(bytes, body);
      channels = u16le(bytes, body + 2);
      sampleRate = int(u32le(bytes, body + 4));
      bitsPerSample = u16le(bytes, body + 14);
      if (audioFormat == 0xFFFE && size >= 40) {
        audioFormat = u16le(bytes, body + 24);
      }
      haveFmt = true;
    } else if (eq4(bytes, offset, "data")) {
      data = bytes.mid(body, size);
      haveData = true;
    }

    offset = body + size + (size % 2);
  }

  if (!haveFmt || !haveData) {
    throw std::runtime_error("WAV missing fmt/data");
  }
  if (audioFormat != 1 && audioFormat != 3) {
    throw std::runtime_error("unsupported WAV format");
  }
  if (bitsPerSample != 16 && bitsPerSample != 24 && bitsPerSample != 32) {
    throw std::runtime_error("unsupported bits");
  }

  PcmBuffer pcm;
  pcm.samples = decodeMono(data, channels, bitsPerSample, audioFormat == 3);
  pcm.sampleRateHz = sampleRate;
  return pcm;
}

}  // namespace tramp
